---
title: "250903_GL_dds_newFilters"
author: "Irika Sinha"
date: "2025-09-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Purpose

The purpose of this script is to analyze rat Cx, Hip, and Spleen tissues for response to LPS and VNS.

## Packages

```{r packages}
CRANpackages <- c("tidyverse", "ggplot2", "ggpubr", 
                  "ggstats","BiocManager","viridis",
                  "gghighlight", "ggrepel","pheatmap","readxl", 
                  "RColorBrewer")
biocPackages <- c("apeglm", "DESeq2","limma", "biomaRt","fgsea")

for(i in CRANpackages){
  if(!(i %in% rownames(installed.packages()))){
    install.packages(pkgs = i)
  }
  library(package = i, character.only = T, quietly = T)
  rm(i)
}

for(i in biocPackages){
  if(!(i %in% rownames(installed.packages()))){
    BiocManager::install(pkgs = i, ask = F)
  }
  library(package = i, character.only = T, quietly = T)
  rm(i)
}

rm(CRANpackages,biocPackages)

```
## Filter samples using QC results from MultiQC
Cortex - 91 has a high rate of duplicates and 85, 87, 96 are more like each other than the others
Hippocampus - 98 has a high rate of duplicates
Spleen - 90 has a very low mean inner distance, potentially indicative of RNA degradation
```{r filterdds}
setwd("/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/")
dds <- readRDS("data/dds_merged.RDS")

## Cortex MultiQC closer look
# QC from multiQC shows G91 has a very high rate of duplicated reads
dups <- read.csv("data/cortex_dups.csv") %>% separate(col = 1, sep = "_", into = c("con","sample","read")) %>%
  mutate(Dups = as.numeric(gsub(pattern = "%", replacement = "", x = Dups)),
         Failed = as.numeric(gsub(pattern = "%", replacement = "", x = Failed)))
ggplot(dups, aes(x = Dups, y = Failed, color = sample, label=sample)) + geom_point(size = 3) + 
  ylim(0,50)+xlim(0,100)+ylab("% Failed Modules in FastQC")+xlab("% Duplicated Reads")+
  theme_classic()+theme(text=element_text(color="black", size = 12))+
  ggrepel::geom_text_repel(max.overlaps = 20,force = 5)

# Samples 85, 87, 96 are more like each other than the others within the same conditions + also have relatively high duplicated reads
pca <- read.csv("data/cortex_star_salmon_deseq2_clustering-plot.csv", row.names = 1)
pca_filt <- apply(pca, c(1, 2), function(x) ifelse(x > 100, 100, x))

pheatmap(as.matrix(pca_filt), treeheight_row = 0, treeheight_col = 0,
         colorRampPalette((brewer.pal(n = 7, name = "RdYlBu")))(100),
         border_color = NA)

## Hippocampus MultiQC shows 98 has a very high rate of duplicated reads
## Spleen MultiQC shows 90 has unexpectedly low inner distance

## Filter dds to remove samples as indicated above
rmSamp <- which(colnames(dds) %in% c("VNS_G91CX","VNS_G85CX","CTRL_G87CX","CTRL_G96CX","LPS_G98HC","CTRL_G90SP"))
dds <- dds[,-rmSamp]
# saveRDS(dds, "data/250903_dds_merged_filtered.RDS")

```

## Filter samples based on GEX & add extra metadata
```{r filterGEX}
setwd("/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/")
dds <- readRDS("data/250903_dds_merged_filtered.RDS")

## Filter out low expression genes - there should be 10 counts in total across all samples
dds <- dds[rowSums(counts(dds)) >= 10,]

## Filter out low expression genes - at least one group should have median expression >= 5
metadata <- data.frame(colData(dds)) %>% mutate(group = paste(treatment, tissue, sep = "_"))
counts <- data.frame(counts(dds))
df <-  data.frame(genes = rownames(counts))

# Create function to calculate median count by group
medExpByGroup <- function(x, countdf, metadatadf){
  # which samples are in group
  samp <- metadata$sample[which(metadata$group %in% x)]
  
  # subset count df
  countdf <- countdf %>% dplyr::select(samp)
  
  # get medians of condition
  medDF <- apply(countdf, 1, function(x) round(median(x),1))
  
  return(medDF)
}

# Apply function
df[,2:10] = lapply(unique(metadata$group), function(x) medExpByGroup(x=x,countdf=counts, metadatadf=metadata))
colnames(df)[2:10] = unique(metadata$group)
df <- df %>% column_to_rownames("genes")

# Filter out genes with median less than 5 counts across all groups
dds <- dds[apply(df, 1, function(x) max(x) > 5),]

## Add metadata
# clean up sample ID
sampleIDs <- dds@colData$sample
sampleIDs <- str_split_i(sampleIDs, "_", 2)
sampleIDs <- gsub("[A-Z]","",sampleIDs)
dds@colData$sampID = sampleIDs
dds$sampID <- factor(dds$sampID, levels = unique(dds$sampID))

# Add nervous tissue subset
tissues = dds@colData$tissue
tissues = gsub("cortex|hippocampus","CNS",tissues)
tissues = gsub("spleen","PNS",tissues)
dds@colData$tissue = tissues

# Add coldata on condition
dds@colData$condition = paste0(dds@colData$treatment,"_", dds@colData$tissue)
dds$condition <- factor(dds$condition, levels = unique(dds$condition))
dds$condition <- relevel(dds$condition, ref = "CTRL_PNS")

# Add design
design(dds) <- formula(~ treatment)

# saveRDS(dds, "/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/data/250903_filtered_annotated_dds.RDS")

# PCA
pcaData <- plotPCA(vst(dds), intgroup=c("tissue", "treatment"), returnData = T)
ggplot(pcaData,aes(x = PC1, y = PC2, color = tissue, shape = treatment)) + geom_point(size = 2, alpha = 0.8)+
  theme_classic()+theme(text=element_text(color="black", size = 12)) + 
  scale_colour_manual(values = c("#5D0D0D", "#111344", "#ADA8B6"))

pcaData <- plotPCA(vst(dds[,which(colData(dds)$tissue == "cortex")]), intgroup=c("treatment"), returnData = T)
ggplot(pcaData,aes(x = PC1, y = PC2, color = treatment)) + geom_point(size = 4, alpha = 1)+
  theme_classic()+theme(text=element_text(color="black", size = 12)) + 
  scale_colour_manual(values = c("#5D0D0D", "#111344", "#ADA8B6")) +
  theme(legend.position = "none")

pcaData <- plotPCA(vst(dds[,which(colData(dds)$tissue == "hippocampus")]), intgroup=c("treatment"), returnData = T)
ggplot(pcaData,aes(x = PC1, y = PC2, color = treatment)) + geom_point(size = 4, alpha = 1)+
  theme_classic()+theme(text=element_text(color="black", size = 12)) + 
  scale_colour_manual(values = c("#5D0D0D", "#111344", "#ADA8B6")) +
  theme(legend.position = "none")

pcaData <- plotPCA(vst(dds[,which(colData(dds)$tissue == "PNS")]), intgroup=c("treatment"), returnData = T)
ggplot(pcaData,aes(x = PC1, y = PC2, color = treatment)) + geom_point(size = 4, alpha = 1)+
  theme_classic()+theme(text=element_text(color="black", size = 12)) + 
  scale_colour_manual(values = c("#5D0D0D", "#111344", "#ADA8B6")) +
  theme(legend.position = "none")


```

## Differential gene expression analysis based on treatment
All organs are grouped together - only difference in treatment is considered

```{r treatmentDGE}
setwd("/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/")
dds <- readRDS("data/250903_filtered_annotated_dds.RDS")

## Use Ensembl database to download gene ENMUS gene names to gene names + subset to protein-coding genes
ensembl <- useEnsembl(biomart = "genes", dataset = "rnorvegicus_gene_ensembl")
mart <- useDataset("rnorvegicus_gene_ensembl", useMart("ensembl"))

gene_names <-getBM(filters= "ensembl_gene_id", 
                   attributes= c("ensembl_gene_id","external_gene_name","gene_biotype"),
                   values=rownames(dds), mart= mart) 
# saveRDS(gene_names, "data/250903_filtered_annotated_dds_geneNames.RDS")


## DESeq2 Analysis: Treatment
dds <- DESeq(dds)
dds_LPS <- lfcShrink(dds, coef="treatment_LPS_vs_CTRL",  type="apeglm") %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  inner_join(gene_names %>% filter(gene_biotype == "protein_coding") %>% 
               dplyr::select(!c("gene_biotype")),
             by = "ensembl_gene_id") %>%
  relocate(external_gene_name, .before = "ensembl_gene_id") %>%
  mutate(external_gene_name = ifelse(external_gene_name == "", ensembl_gene_id, external_gene_name))

dds_VNS <- lfcShrink(dds, coef="treatment_VNS_vs_CTRL",  type="apeglm") %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  inner_join(gene_names %>% filter(gene_biotype == "protein_coding") %>% 
               dplyr::select(!c("gene_biotype")),
             by = "ensembl_gene_id") %>%
  relocate(external_gene_name, .before = "ensembl_gene_id") %>%
  mutate(external_gene_name = ifelse(external_gene_name == "", ensembl_gene_id, external_gene_name))

# merge dfs
mergedf <- dds_LPS[,c("external_gene_name", "ensembl_gene_id", "log2FoldChange", "padj")] %>%
  full_join(dds_VNS[,c("external_gene_name", "ensembl_gene_id","log2FoldChange", "padj")], 
            by = c("external_gene_name","ensembl_gene_id"),suffix = c("_LPS","_VNS"))
# saveRDS(mergedf,"/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/output/250903_mergedTreatment_df.RDS")

```

Visualize top genes across the different conditions
```{r treatmentDGE_vis}
setwd("/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/")
mergedf <- readRDS("output/250903_mergedTreatment_df.RDS")

# Filter by expression (log2(0.75) = -0.42, log2(1.25) = 0.32)
df <- mergedf 
df$log2fcMax = apply(df[,c("log2FoldChange_LPS","log2FoldChange_VNS")], 1, max)
df$log2fcmin = apply(df[,c("log2FoldChange_LPS","log2FoldChange_VNS")], 1, min)
df$log2fcdiff = apply(df[,c("log2FoldChange_LPS","log2FoldChange_VNS")], 1, var )
df_filter <- df %>% filter(case_when(log2fcMax < 0.32 ~ log2fcmin < -0.42, 
                   log2fcmin > -0.42 ~ log2fcMax > 0.32,
                   log2fcmin < -0.42 ~ log2fcMax > 0.32))

# Top 30 up- and down-regulated
df_top <- df_filter %>% slice_max(log2fcMax, n=30) %>%
  pivot_longer(cols=c("log2FoldChange_LPS","log2FoldChange_VNS"), 
               names_to = "condition_lfc", names_prefix = "log2FoldChange_", 
               values_to = "log2FC") %>%
    pivot_longer(cols=c("padj_LPS","padj_VNS"), 
               names_to = "condition_padj", names_prefix = "padj_", 
               values_to = "padj") %>% 
  mutate(keep = "N", keep = ifelse(condition_lfc == condition_padj, "Y","N")) %>%
  filter(keep == "Y") %>% dplyr::select(-c("keep","condition_padj"))

df_tail <- df_filter %>% slice_min(log2fcmin, n=30) %>%
  pivot_longer(cols=c("log2FoldChange_LPS","log2FoldChange_VNS"), 
               names_to = "condition_lfc", names_prefix = "log2FoldChange_", 
               values_to = "log2FC") %>%
    pivot_longer(cols=c("padj_LPS","padj_VNS"), 
               names_to = "condition_padj", names_prefix = "padj_", 
               values_to = "padj") %>% 
  mutate(keep = "N", keep = ifelse(condition_lfc == condition_padj, "Y","N")) %>%
  filter(keep == "Y") %>% dplyr::select(-c("keep","condition_padj")) 
order <- rev(unique(c((df_top %>% arrange(-log2fcMax))$external_gene_name, 
                      (df_tail %>% arrange(-log2fcmin))$external_gene_name)))

df_plot <- rbind(df_top,df_tail) %>% 
  mutate(external_gene_name = factor(external_gene_name, levels = order)) %>% 
  mutate(pLabel = ifelse(padj < 0.001, "***",ifelse(padj < 0.01, "**",ifelse(padj < 0.05, "*",""))), 
         labelC = ifelse(round(log2FC,0) > 0, "black", "white"))


ggplot(df_plot, aes(x=external_gene_name, y = condition_lfc, 
                    fill = log2FC, label = pLabel)) +
  geom_tile(color = "white") +theme_classic()+
  xlab("")+ylab("") + 
  theme(text =  element_text(color = "black", size = 12),
        axis.line = element_blank(), 
        axis.ticks = element_blank())+
  scale_y_discrete(expand = c(0,0))+
  geom_text(color = df_plot$labelC, size = 4)+
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust=1))+
  scale_fill_viridis_c(limits = c(-3, 8),option="A")

# Filter by difference (at least 2 fold)
df_filter <- df %>% filter(log2fcdiff > 1)
order <- rev(unique(df_filter %>% arrange(-log2FoldChange_LPS))$external_gene_name)

df_plot <- df_filter %>% 
  mutate(external_gene_name = factor(external_gene_name, levels = order))  %>%
  pivot_longer(cols=c("log2FoldChange_LPS","log2FoldChange_VNS"), 
               names_to = "condition_lfc", names_prefix = "log2FoldChange_", 
               values_to = "log2FC") %>%
    pivot_longer(cols=c("padj_LPS","padj_VNS"), 
               names_to = "condition_padj", names_prefix = "padj_", 
               values_to = "padj") %>% 
  mutate(keep = "N", keep = ifelse(condition_lfc == condition_padj, "Y","N")) %>%
  filter(keep == "Y") %>% dplyr::select(-c("keep","condition_padj")) %>% 
  mutate(pLabel = ifelse(padj < 0.001, "***",ifelse(padj < 0.01, "**",ifelse(padj < 0.05, "*",""))), 
         labelC = ifelse(round(log2FC,0) > 0, "black", "white")) 

ggplot(df_plot, aes(x=external_gene_name, y = condition_lfc, 
                    fill = log2FC, label = pLabel)) +
  geom_tile(color = "white") +theme_classic()+
  xlab("")+ylab("") + 
  theme(text =  element_text(color = "black", size = 12),
        axis.line = element_blank(), 
        axis.ticks = element_blank())+
  scale_y_discrete(expand = c(0,0))+
  geom_text(color = df_plot$labelC, size = 4)+
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust=1))+
  scale_fill_viridis_c(limits = c(-3, 6),option="A")


```

## Differential gene expression analysis compared to LPS as set-point
All organs are grouped together - only difference in treatment is considered

```{r treatmentDGE}
setwd("/Users/irika/Library/CloudStorage/OneDrive-JohnsHopkins/WongLing/Others/250601_GL/2025_GL_RatSeq/")
dds <- readRDS("data/250903_filtered_annotated_dds.RDS")

## Use Ensembl database to download gene ENMUS gene names to gene names + subset to protein-coding genes
gene_names <- readRDS("data/250903_filtered_annotated_dds_geneNames.RDS")

## DESeq2 Analysis: Treatment
dds$treatment <- relevel(dds$treatment, ref = "LPS")
dds <- DESeq(dds)
dds_VNS <- lfcShrink(dds, coef="treatment_VNS_vs_LPS",  type="apeglm") %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  inner_join(gene_names %>% filter(gene_biotype == "protein_coding") %>% 
               dplyr::select(!c("gene_biotype")),
             by = "ensembl_gene_id") %>%
  relocate(external_gene_name, .before = "ensembl_gene_id") %>%
  mutate(external_gene_name = ifelse(external_gene_name == "", ensembl_gene_id, external_gene_name))

dds_Ctrl <- lfcShrink(dds, coef="treatment_CTRL_vs_LPS",  type="apeglm") %>% as.data.frame() %>%
  rownames_to_column("ensembl_gene_id") %>%
  inner_join(gene_names %>% filter(gene_biotype == "protein_coding") %>% 
               dplyr::select(!c("gene_biotype")),
             by = "ensembl_gene_id") %>%
  relocate(external_gene_name, .before = "ensembl_gene_id") %>%
  mutate(external_gene_name = ifelse(external_gene_name == "", ensembl_gene_id, external_gene_name))

# merge dfs
mergedf <- dds_VNS[,c("external_gene_name", "ensembl_gene_id", "log2FoldChange", "padj")] %>%
  full_join(dds_Ctrl[,c("external_gene_name", "ensembl_gene_id","log2FoldChange", "padj")], 
            by = c("external_gene_name","ensembl_gene_id"),suffix = c("_VNS","_Ctrl"))
# saveRDS(mergedf,"output/250903_mergedTreatment_VNSref_df.RDS")

```

There aren't any significant differences between VNS and LPS treatments still.